internal/states/encryption/flow/interface.go

```diff
@@ -0,0 +1,102 @@
+// Package flow contains the top-level flow for client-side state encryption and the interfaces an encryption
+// or key derivation method need to implement.
+package flow
+
+import "github.com/opentofu/opentofu/internal/states/encryption/encryptionconfig"
+
+type ConfigurationSource string
+
+const (
+	ConfigurationSourceEnvDefault ConfigurationSource = "env-default"
+	ConfigurationSourceCode       ConfigurationSource = "code"
+	ConfigurationSourceEnv        ConfigurationSource = "env"
+)
+
+// Flow represents the top-level state or plan encryption/decryption flow
+// for a particular encryption configuration.
+//
+// Instances of Flow are kept in an internal cache per configuration key.
+// Unless you are writing tests, you should not create them in your code.
+//
+// See also encryption.Instance().
+type Flow interface {
+	// DecryptState decrypts encrypted state.
+	//
+	// If the state is not actually encrypted, it will be returned
+	// as-is. DecryptState will first try decryption using the state
+	// encryption configuration. If this fails, it tries the decryption
+	// fallback configuration. If neither produces a valid result it fails.
+	//
+	// payload must be a json document passed in as a []byte.
+	// This can be encrypted or unencrypted state.
+	//
+	// If no error is returned, then the first return value will always
+	// be a json document, possibly the same one passed in as payload
+	// if the payload was not actually encrypted.
+	DecryptState(payload []byte) ([]byte, error)
+
+	// EncryptState encrypts plaintext state.
+	//
+	// There are special configurations that will not actually encrypt
+	// state. This happens when you only configure a decryption fallback,
+	// but not encryption. This is not an error.
+	//
+	// state is a json document passed in as a []byte.
```

Comment:
I'm debating with myself about this. Theoretically, someone could pass in any JSON to this function to undefined results. I wonder if the format that needs to be passed in should be explicitly documented here and then enforced. This may be especially important if future developers change the format of the state file or something along those lines as this may break the encryption.

I'm thinking about cases where new fields are added, etc.

Reply:
I am hoping we can keep the currently planned encryption methods (full, partial = all the string leaf nodes) agnostic of the rest of the json contents (unless someone introduces "encryption" at top level, which would cause be a collision).

I'll add documentation that the json can't have a key "encryption". That at least should cause an error.

Reply:
I wonder if that special key could be made configurable. That would make the API more explicit about what it does.

Reply:
Good idea. Makes it explicitly visible in the code rather than just the comments. I'll do that.